### ISSUE TITLE
A feature to add custom comments to PR as Issue comments

### ISSUE THREAD

**Issue by neozonka** [jenkinsci/ghprb-plugin#356]
It would be awesome if a post-build step can be added which can post Issue comments to a PR once the test is completed and this comment can be custom comment unlike the jenkins URL comment feature.

**Comment by reportingsjr** [comment]
Please, please, please! This would be phenomenal to have.

The hack I am currently using for this is as follows:
  curl -i -H "Authorization: token githubtokenhere" \
       -H "Content-Type: application/json" \
       -d "{\"body\":`python -c 'import json,sys; print(json.dumps(\"Please fix these issues:\n\" + open(sys.argv[1]).read()))' result.txt`}" \
       https://api.github.com/repos/reportingsjr/kicad-library/issues/$ghprbPullId/comments

However this displays the github token in the log (huge security issue) and is hideous.

**Comment by benpatterson** [comment]
I would think this could be done with the downstream job pattern described in the README nowadays?

**Comment by reportingsjr** [comment]
I see where you can update the commit status, but not add a new comment. Am I missing something?

**Comment by benpatterson** [comment]
@reportingsjr I was thinking specifically of "The job has passed" as part of this line in jobDSL: `completedStatus('SUCCESS', "The job has passed")` 

Is that what you folks have in mind? If not, why not?

**Comment by reportingsjr** [comment]
Isn't that just updating the status message of the commit?

If so then that is not what I want. I want a full comment. It's more obvious to people and I can embed pictures and formatted text.

**Comment by benpatterson** [comment]
@reportingsjr Ohhhh do you mean a GitHub comment? Like the bot user will come back to the PR and give some information?

**Comment by reportingsjr** [comment]
Exactly! Isn't the title of this "add custom comment to PR"? haha :)

**Comment by benpatterson** [comment]
Haha I think I was looking at it upside-down or something; thanks for (re)clarifying.

**Comment by karlhorky** [comment]
Is this not the same feature that `ghprb` apparently already has (at least the upstream)? https://github.com/janinko/ghprb/issues/293

That feature shows up like this for me in the job configuration:

![bildschirmfoto 2016-10-15 um 14 48 48](https://cloud.githubusercontent.com/assets/1935696/19410223/8c08bf68-92e6-11e6-9f68-0cc1760d55b3.png)

This results in the following messages in our pull requests:

![bildschirmfoto 2016-10-15 um 15 47 32](https://cloud.githubusercontent.com/assets/1935696/19410510/bffe1bbc-92ee-11e6-95de-6c3e9e24cdf9.png)

**Comment by reportingsjr** [comment]
No, that does not let you put a custom comment for every run. I want to be able to tell people in a comment the errors generated from a pull request.

**Comment by karlhorky** [comment]
Ah yes, I would like this too! Was this the point of the original issue here though?

**Comment by andrewfraley** [comment]
The method mentioned by @karlhorky does let you add comments to the pull request.  It's confusing because it looks like you're adding the same thing twice, but to get a comment you have to add a separate "Build Status Message".  In triggers, under (and outside) of "Update commit status during build", click add->"Build Status Messages".  Then add "Build result messages".

There seems to be a bug where these comments don't expand Jenkins variables, however, so I'll open a separate issue for that.

**Comment by wontonst** [comment]
The readme says ```You can extend the standard build comment message on github creating a comment file from shell ``` but I don't see any instructions or Jenkins steps on how this is actually implemented.

**Comment by bjoernhaeuser** [comment]
The plugin supports settings a test one liner for the context.

I assume this solves some part of the problem?

**Comment by kthacker7** [comment]
Nope this problem is still unsolved. I want the comment to be the results of the number of unit tests passed vs failed and static text doesn't allow me to do that.

**Comment by grahamwhaley** [comment]
the ghprb does have a 'comment file' option. It is deeply buried in the setup. I've not tried it yet (rsn), but to find it you have to go to the config section of ghprb on your build and then follow:
Trigger Setup -> Add[v] (the second add - the dropdown one)
and there you should see the `comment file` option.

**Comment by claudio-viola** [comment]
For anyone who still gets here.
Provided that 
/var/lib/jenkins/jobs/ 
is your jenkins jobs path
add 
```
/var/lib/jenkins/jobs/${JOB_NAME}/builds/${BUILD_ID}/github 
```
to your Comment File Path as explained by grahamwhaley.

in your shell compilation add this

```
BUILD_FOLDER=/var/lib/jenkins/jobs/${JOB_NAME}/builds/${BUILD_ID}
GITHUB_COMMENT_FILE=$BUILD_FOLDER/github
rm -rf $GITHUB_COMMENT_FILE
***** YOUR LAST SUCCESS COMMAND HERE *******
if [ $? -ne 0 ]; then
  # PUT LATEST 50 lines of error logs or more or whatever you want to put in the issue comment
  echo -e "Tests failed" >> $GITHUB_COMMENT_FILE
  echo -e "Full logs available at $BUILD_URL" >> $GITHUB_COMMENT_FILE
  tail -n 50 $BUILD_FOLDER/log >> $GITHUB_COMMENT_FILE
else
  echo -e "Tests pased" >> $GITHUB_COMMENT_FILE
fi

```
